<p align = "center">
<img width="250" src="https://github.com/marciow0205/GeneSelectorForValidation/blob/087abbb4b8dd3f2f4a9e1a4f82f21fab2c4e133e/image/logo2.png" alt="GSV-logo"/></p>
<h1 align="center">Gene Selector for Validation</h1>
<!-- Badges -->

[![Python Version](https://img.shields.io/badge/Python-3.6%2B-blue)](https://www.python.org/downloads/)
[![Platform](https://img.shields.io/badge/Platform-Windows%2010-blue)](https://www.microsoft.com/en-us/windows/windows-10-features)
<img alt="GitHub repo size" src="https://img.shields.io/github/repo-size/marciow0205/TCC">
<img alt="GitHub last commit" src="https://img.shields.io/github/last-commit/marciow0205/TCC">
<p align="center"><i>Uma ferramenta capaz de indicar candidatos a genes de referência e de validação para o seu RT-qPCR.</i></p>

## Index

- [Description](#description)
- [Installation](#installation)
- [How to use](#how-to-use)
- [Accepted file types](#accepted-files-types)
- [Methodology](#methodology)
- [How to contribute](#how-to-contribute)

## Description
The Gene Selector for Validation (GSV) is a program developed in the Python programming language. The algorithm follows a filtering-based methodology, using Transcripts Per Million (TPM) values to compare gene expression across bulk RNA-seq samples.

The GSV groups transcriptome quantification tables, relating the TPM value of each gene according to the analyzed RNA-seq library in a DataFrame. Then, established criteria are applied to remove all genes that do not meet the requirements. Finally, a file is returned with a table indicating which genes are the most stable and which are the most variable across all libraries.

## Installation

To use the Gene Selector for Validation (GSV), it is not necessary to install Python or other dependencies on your machine, as the program has already been compiled into an executable file (.exe). Simply download the file from the GitHub repository and run it on your Windows 10 computer.

  1. Access the main page of the GSV repository on GitHub: https://github.com/rdmesquita/GSV

  2. Download the latest available release.

  3. Extract the file to a folder of your choice. Make sure the "image" folder is present in the same directory as the executable file.

  4. Double-click on the "GeneSelectorforValidation.exe" file to run the program.

That's it! The program is now ready to use.

## How to use
To use the Gene Selector for Validation (GSV), follow these steps:

  1. Select the RNA-Seq data to be analyzed.
  
  2. Configure the program according to the type of input file and set filter values.
    * The GSV accepts files with extensions .csv, .xls, .xlsx, and .sf, the latter being an extension generated by the Salmon software. When the user selects the input file, the program requests additional information depending on the file format. This information is requested during the program configuration step.
    * For .csv, .xls, and .xlsx formats, only one file is accepted, which must contain a table relating the TPM of each gene according to the RNA-seq library used in the analysis. In this format, replicas are not accepted, and if there are any, the user must average them before using the program.
    * In the .sf format, one file is accepted for each library. If there are replicas, they must be sent to the program in order.
    
  3. Click on "Analyze" to perform the analysis.

## System requirements
The program is only compatible with the Windows 10 operating system. There are no minimum hardware requirements. The program is already compiled into an .exe file, and the user does not need to have Python or other dependencies installed on the machine.

### Observations
The "image" folder must be present in the same directory as the executable file.

  
## 3.1 - Choosing input data
  The .csv, .xls, and .xlsx data should be a single table containing genes and their TPM values in each library, preferably as shown in the example below
  <table align = "center">
    <thead>
        <th style="width: 25%;">
            Gene
        </th>
        <th style="width: 25%;">
            Library 1
        </th>
        <th style="width: 25%;">
            Library 2
        </th>
        <th style="width: 25%;">
           Library 3
        </th>
    </thead>
    <tbody>
        <tr>
            <td>
                Gene X
            </td>
            <td>
                TPM
            </td>
            <td>
                TPM
            </td>
            <td>
                TPM
            </td>
        </tr>
        <tr>
            <td>
                 Gene A
            </td>
            <td>
                TPM
            </td>
            <td>
                TPM
            </td>
            <td>
                TPM
            </td>
        </tr>
        <tr>
            <td>
                  Gene B
            </td>
            <td>
                TPM
            </td>
            <td>
                TPM
            </td>
            <td>
                TPM
            </td>
        </tr>
        <tr>
            <td>
                Gene C
            </td>
            <td>
                TPM
            </td>
            <td>
                TPM
            </td>
            <td>
                TPM
            </td>
        </tr>
    </tbody>
</table>


The files derived from the Salmon software (.sf extension) have the following format:

<b>PS: Mere illustrative numbers</b>

<table align = "center">
    <thead>
        <th style="width: 25%;">
            Name
        </th>
        <th style="width: 25%;">
            Length
        </th>
        <th style="width: 25%;">
            EffectiveLength
        </th>
        <th style="width: 25%;">
            TPM
        </th>
        <th style="width: 25%;">
            NumReads
        </th>
    </thead>
    <tbody>
        <tr>
            <td>
                Gene X
            </td>
            <td>
                1031
            </td>
            <td>
                781.000
            </td>
            <td>
                4.852.151
            </td>
            <td>
                107.000
            </td>
        </tr>
        <tr>
            <td>
                 Gene A
            </td>
            <td>
                1437
            </td>
            <td>
                1.187.000
            </td>
            <td>
                0.0000
            </td>
            <td>
                0.000
            </td>
        </tr>
        <tr>
            <td>
                  Gene B
            </td>
            <td>
                1430
            </td>
            <td>
                1.180.000
            </td>
            <td>
                0.0000
            </td>
            <td>
                0.0000
            </td>
        </tr>
        <tr>
            <td>
                Gene C
            </td>
            <td>
                1634
            </td>
            <td>
                993.000
            </td>
            <td>
                0.4635
            </td>
            <td>
                16.664
            </td>
        </tr>
    </tbody>
</table>

In this case, the GSV is capable of extracting only the necessary information for the analysis, the Name column and the TPM column

## 3.2 - Configuring the program

Each file extension chosen must be configured differently. When clicking on the "2 - Set Files..." button, the file extension must be selected. Depending on the file type, GSV will ask for different information.

> Text files (.csv) - 
  When selecting this file type, GSV will request two pieces of information such as the column name where the genes are located and the separator character for that file (it is always recommended to use ";" as a separator).

> Tabular files (.xls and .xlsx) - 
  Only the column name of the genes is requested in these files.

> Salmon files (.sf) - 
  Here, in addition to the gene column names, the TPM value column name and the number of replicates are also requested.

After choosing the file extension of the uploaded files and providing the requested information, the user must click "To apply" for the information to be sent.
  
## 3.3 - Filter Configuration
  In "3 - Set Filters...", the user can change the default filter values, but it is HIGHLY recommended that the user keeps the default values. In this window, it is also possible to read what each filter means through an informative pop-up menu that appears when hovering over each one.
  
## 3.4 - Results
  At the end of the analysis, two new windows appear, one with results related to the candidate reference genes that show high stability and expression, and the second window related to the validation genes that show high expression and high variation in expression.
This result can be saved in .xlsx, .xls, and .txt so that it can be analyzed by the user.
